### ISSUE TITLE
[BUG] Window activation issues for x-callback handlers (e.g., zotero://)

### ISSUE THREAD

**Issue by jimvine** [Zettlr/Zettlr#283]
It would be great if the zotero: protocol were recognised, so that I could insert functioning zotero://select links into my notes.

Briefly, I can use Zutilo to obtain the Zotero Select link to an item in my Zotero database.

I can add a link to it in Zettlr using the format `[zotero://select/items/1_12345678](zotero://select/items/1_12345678)`.

This is recognised as a link and if I hold Ctrl my pointer turns to a hyperlink-hand.

However, ctrl-clicking the link does not open it.

I believe that it might be as simple as adding zotero to the list of protocols in line 22 of 
`/source/renderer/assets/codemirror/zettlr-plugin-render-links.js`.

**Comment by nathanlesage** [comment]
The file you're referring to will only help in rendering standalone `zotero`-links. Additionally, I pull the regular expression right from the underlying Markdown mode, so I don't like the idea of un-syncing the detection of links from the highlighting mode.

Zettlr will try to agnostically open _any_ file using electron's `shell.openExternal`, so it seems Zotero-links are not registered on your system. On my machine (macOS Mojave 10.14.5) Zotero links actually open Zotero.

A little test: In your webbrowser, try opening one of the zotero-links, and if this doesn't work, then it seems to be a problem with the registration of the x-callback-URL of Zotero.

**Comment by jimvine** [comment]
Fair enough about not messing with dependencies in the regex.

If I paste the zotero:select URL into the URL bar on my browser and press enter, it correctly takes me to the expected item in Zotero. (Windows 10, Chrome.)

I can try on a different machine later or tomorrow to see how Zotero and Zettlr behave there.

**Comment by nathanlesage** [comment]
Yep, but if it works on your machine using a browser, then it does indicate a bug in the Electron framework -- are you using 1.3 or 1.4-beta? On the latter at least it works, I haven't tried it on 1.3

**Comment by jimvine** [comment]
**TL;DR: Zettlr does seem to open zotero://select links but doing so does not always give the Zotero application the focus.**

I am now trying this on a different machine (still Windows 10 with Chrome as the default browser) and I think I may have been wrong in thinking that Zettlr does nothing when clicking a zotero://select link. It does do something, but not always as much as I expected. It may well have been doing this on the other machine without me noticing.

When I paste a zotero://select URL into a browser bar it brings Zotero to the fore and goes to that item.

Same thing when I test with a zotero://select link on a test HTML page:

```
<html>
    <body>
        <p><a href="zotero://select/items/1_12345678">Zotero select test</a></p>
    </body>
</html>
```

Using a link in Chrome:

* If Zotero is closed, it gets launched, brought to the fore and goes to that item.
* If Zotero is open but minimised, it gets maximised, brought to the fore and goes to that item.
* If Zotero is open, maximised but behind other windows, it gets brought to the fore and goes to that item.

Using a link in Zettlr:

* If Zotero is closed, it gets launched, brought to the fore and goes to that item.
* If Zotero is open but minimised, it gets maximised, brought to the fore and goes to that item.
* If Zotero is open, maximised but behind other windows, it goes to that item **but stays in the background and *is not* brought to the fore.**

I will double-check when I am back at the other system, but I suspect that the lack of visual feedback (with the Zotero window not popping up) may have caused me to act too hastily in assuming 'nothing had happened'.

Now, for my purposes --- and I suspect in many cases --- it would be great if opening a link also consistently caused the link handler to gain focus.

**Comment by jimvine** [comment]
And now, on further investigation I think that this might be completely unrelated to Zotero.

I have just tried opening normal web links in Zettlr using the states above. When opening an `https://` link from Zettlr:

* If Chrome is minimised it gets maximised and gains focus.
* But if Chrome is maximised but hidden behind other windows then the link opens but Chrome stays in the background.

**Comment by jimvine** [comment]
> Yep, but if it works on your machine using a browser, then it does indicate a bug in the Electron framework -- are you using 1.3 or 1.4-beta? On the latter at least it works, I haven't tried it on 1.3

Sorry - hadn't spotted this question that you were posting at the same time as mine earlier. For the record I'm using 1.3 on both machines.

**Comment by nathanlesage** [comment]
Thanks for the feedback!

I strongly suggest this to be a small bug in the Electron framework, because I've been observing the problem with target apps not receiving focus — apparently it's an option that has never been declared mandatory, as to this documentation snippet:

![grafik](https://user-images.githubusercontent.com/17251683/62192613-36e4fc00-b376-11e9-82c2-a48d6fe28d7f.png)

Additionally, there seems to have been a bug fixing concerning this option on some more operating systems, about a month ago: https://github.com/electron/electron/pull/18721/files

I'll test out how passing the option works out in 1.4.0-beta.2!

**Comment by nathanlesage** [comment]
I'll close this issue as the handlers themselves worked all the time, and maybe on Windows they now fixed the delay!
